### ISSUE TITLE
Error 500 in production

### ISSUE THREAD

**Issue by xaitec** [adamghill/django-unicorn#408]
Every thing runs fine in development but when I try to run on the production server I get this message. I will say that some calls work about 20% of the time. Not sure what else to do or what info to give as it works well in development.

For production I use a docker container with nginx as a proxy. 

Can unicorn assets be served by whitenoise?  Would I need to have nginx handle serving this url?

![image](https://user-images.githubusercontent.com/678838/169033166-29d85301-8258-4cd6-a6af-78eca137eb69.png)

**Comment by adamghill** [comment]
It's curious that things work 20% of the time. You are showing the JS console error, but can show the Python error log on the server (or do you use something like Sentry, Datadog, etc.)? That will give you the Python traceback and might give us some more clues.

**Comment by xaitec** [comment]
This is the message that was returned

![image](https://user-images.githubusercontent.com/678838/169523322-80ab0b92-1cea-4dbc-85e5-5bac77d26643.png)

**Comment by adamghill** [comment]
Hmm, I've never seen that happen only in production before. Is `employee_services.components.employee_disabilities.EmployeeDisabilitiesView` the correct class? Do you have different settings for production than local dev? Maybe `UNICORN` [app settings](https://www.django-unicorn.com/docs/settings/#apps) are different per environment?

**Comment by xaitec** [comment]
Yes it's the correct class. I have not looked into the unicorn app
settings yet, I like to use default settings unless necessary.

**Comment by xaitec** [comment]
I would like to add that I do have django-htmx in my project, as i was planning to convert components to unicorn.

**Comment by adamghill** [comment]
Yeah, I wouldn't expect `htmx` to interfere with `Unicorn`, but I'm glad you tried just in case. That error is raised at https://github.com/adamghill/django-unicorn/blob/main/django_unicorn/components/unicorn_view.py#L867 which should only happen if the `EmployeeDisabilitiesView` class could not be instantiated. I'm not sure how else to debug this problem that only happens 20% of time and only in production, but I'd love to help out however I can.

**Comment by xaitec** [comment]
Just a silly question but is https a requirement?

**Comment by xaitec** [comment]
I understand that. I recently tried again as I upgraded to Django 4. The application in now not public and only used on the LAN. The first action to the unicorn component works, but any other action gives the same browser error as before. I will investigate more on my down time and give the feedback.


On a side note I really like the project and wondered if there was anyway I can help (non financial for now)

**Comment by adamghill** [comment]
I'm sorry it's still not working. :/ I'll try to dig into this again when I have time. 

Since this is just a side project of mine, any help is appreciated! PRs for features/bug fixes, documentation updates, adding better test coverage, etc. :)

**Comment by nickpolet** [comment]
Does this having anything to do with Gunicorn at all?

When I start my application with something like `gunicorn --workers=4  app.wsgi` a lot of requests fail, but some go through. 

If I start gunicorn with only 1 worker, then everything works perfectly.

Also, if I start it with 'gunicorn --workers=4 --worker-class=gevent app.wsgi' and use gevent, then again, everything works 100% of the time.

@adamghill Is this expected behaviour?

**Comment by adamghill** [comment]
Hmm, the number of workers shouldn't matter. My [gunicorn.conf](https://github.com/adamghill/django-unicorn.com/blob/main/gunicorn.conf) for django-unicorn.com.

One thing I didn't ask above: what are you cache settings in production? `Unicorn` uses cache to store component information and I'm wondering if you are using local memory the cache would fail for some of the workers, but would work for one of them perhaps? I always use `redis` in production, so maybe I never saw this before?

**Comment by xaitec** [comment]
I was using in memory cache, 3 workers.

**Comment by adamghill** [comment]
If you are able to try it with a filesystem cache (if you only have one server running) or memcache/redis if you have more than one server that would be helpful. I can also come up with some fallback mechanisms if caching is not available.

**Comment by xaitec** [comment]
Ok, I'll check it out. I really think this can be the real issue

**Comment by regoawt** [comment]
FWIW I had the same issue, and changing from a memory cache to a Memcached instance has seemed to fix it.

Thanks @adamghill for the pointer with this!

**Comment by thewebcat** [comment]
Guys and @adamghill I'd be very grateful for the advice. I had the save issue in production. I started app with gunicorn --workers=4  app.wsgi and every 1-2 requests was with 500 error: "ComponentClassLoadError: The component class could not be loaded: 'NoneType' object has no attribute 'user'". So it was expecting an 'object.user'.
And only for components where i pass some params like {% unicorn 'component' model=object %}.
--
After reading this post i changed django caching from memory to filesystem and everything started working.
But if I leave the tab with the application open, after some time the problem with 500 appears again and  is fixed only by reloading the page.

**Comment by xaitec** [comment]
So from what I see in the [django docs](https://docs.djangoproject.com/en/5.0/topics/cache/#cache-arguments), the default timeout is 300 (5 mins). so mabe you may want to change that for you needs.

**Comment by adamghill** [comment]
If you only have one server than filesystem caching should work ok. But, you might need to bump up the default timeout like @xaitec mentions. I would still like to handle this more gracefully, but I use the cache to store server state between requests.

One question: would it be a good idea to be able to use a database for this functionality instead of only using the cache?

**Comment by xaitec** [comment]
How will that affect performance? Is it just a json field? Is there some
way to refresh the component instead? I like the idea for persistance but
we need to define how cleanup is done

**Comment by thewebcat** [comment]
> If you only have one server than filesystem caching should work ok. But, you might need to bump up the default timeout like @xaitec mentions. I would still like to handle this more gracefully, but I use the cache to store server state between requests.
> 
> One question: would it be a good idea to be able to use a database for this functionality instead of only using the cache?

For sure i'am using database for store my data, i my case "object" is model instance, that seems stored in cache like state of component. So thats why i throw theese error 'NoneType' object has no attribute 'user'.


> So from what I see in the [django docs](https://docs.djangoproject.com/en/5.0/topics/cache/#cache-arguments), the default timeout is 300 (5 mins). so mabe you may want to change that for you needs.

Thanks for your replies guys. I may not quite understand the logic. I mean the state of the component is stored in the cache and has some ttl. So in any case after some time (set in ttl) the component will lose its state and there will be an error?

**Comment by xaitec** [comment]
Correct, there will be an error after the timeout.

What he was mentioning for the database was opinions on if to save the
state in the db instead of the cache so that the state will persist.

**Comment by pa-hqt** [comment]
> So from what I see in the [django docs](https://docs.djangoproject.com/en/5.0/topics/cache/#cache-arguments), the default timeout is 300 (5 mins). so mabe you may want to change that for you needs.

I also observe the ComponentClassLoadError "The component class could not be loaded: 'NoneType' object has no attribute" from time to time and am grateful for the hint to the timeout setting. I will give that a try.

> If you only have one server than filesystem caching should work ok. But, you might need to bump up the default timeout like [@xaitec](https://github.com/xaitec) mentions. I would still like to handle this more gracefully, but I use the cache to store server state between requests.
> 
> One question: would it be a good idea to be able to use a database for this functionality instead of only using the cache?

I would really appreciate it if this error could be treated differently than with a crash.

But I am not shure if a database solution is needed from your side.
Instead one could define a special cache for unicorn that uses the database backend and if needed turns off the timeout.
Wouldn't that achieve the same goal?

Precondition:
django-unicorn allows to define a custom cache-alias that will be used to cache the components. currently the documentation says: _"The alias to use for caching. **Only used by the experimental serialization** of requests for now"_ (https://www.django-unicorn.com/docs/settings/#cache-alias). I am not sure, what that means: is it used for caching components or not?

My **untested** idea:
```python 
# settings.py

# Caching
# https://docs.djangoproject.com/en/4.1/topics/cache/
CACHES = {
    "unicorn": {
        "BACKEND": "django.core.cache.backends.db.DatabaseCache",
        "TIMEOUT": None,  # disable = never expires
    },
}

# 3rd party plugin (django-unicorn)
# https://www.django-unicorn.com/docs/settings/
UNICORN = {
    # The alias to use for caching. Only used by the experimental serialization of requests for now. Defaults to "default"
    "CACHE_ALIAS": "unicorn",
}
```
